### PR TITLE
[libc] add spin lock family functions

### PR DESCRIPTION
**Pull request by SchrodingerZhu**
This PR:
- add entrypoints for `pthread_spin_*`
- additionally, the fixes a typo that has been disabling lock related tests

### CODE REVIEW

File: libc/src/pthread/pthread_spin_destroy.cpp [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+static_assert(sizeof(pthread_spinlock_t::__lockword) == sizeof(SpinLock) &&
+                  alignof(decltype(pthread_spinlock_t::__lockword)) ==
+                      alignof(SpinLock),
+              "pthread_spinlock_t::__lockword and SpinLock must be of the same "
+              "size and alignment");
+
+LLVM_LIBC_FUNCTION(int, pthread_spin_destroy,
+                   ([[maybe_unused]] pthread_spinlock_t * lock)) {
+  auto *spin_lock = reinterpret_cast<SpinLock *>(&lock->__lockword);
```
Comment (lntue) [inline libc/src/pthread/pthread_spin_destroy.cpp:23]: should we check if `lock == nullptr` first before accessing `__lockword`?  And after that can `spinlock` be assumed to be non-null?

File: libc/src/pthread/pthread_spin_init.cpp [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+                  alignof(decltype(pthread_spinlock_t::__lockword)) ==
+                      alignof(SpinLock),
+              "pthread_spinlock_t::__lockword and SpinLock must be of the same "
+              "size and alignment");
+
+LLVM_LIBC_FUNCTION(int, pthread_spin_init,
+                   (pthread_spinlock_t * lock, [[maybe_unused]] int pshared)) {
+  // The spin lock here is a simple atomic flag, so we don't need to do any
+  // special handling for pshared.
+  ::new (&lock->__lockword) SpinLock();
```
Comment (lntue) [inline libc/src/pthread/pthread_spin_init.cpp:26]: Do we need to check whether `lock == nullptr` first?

File: libc/src/pthread/pthread_spin_destroy.cpp (hunk 1 above)
Comment (SchrodingerZhu) [reply libc/src/pthread/pthread_spin_destroy.cpp:23]: ouch, I wrote sth silly. Will correcting it.

**Comment by SchrodingerZhu** [comment]
@lntue @jhuber6 I have updated this to follow up latest tid cache mechanism. Please take a look. Thank you!

File: libc/src/pthread/pthread_spin_destroy.h [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_LIBC_SRC_PTHREAD_PTHREAD_SPIN_DESTROY_H
+#define LLVM_LIBC_SRC_PTHREAD_PTHREAD_SPIN_DESTROY_H
+
+#include "src/__support/macros/config.h"
+#include <pthread.h>
```
Comment (lntue) [inline libc/src/pthread/pthread_spin_destroy.h:13]: Do we have any issues created for pthread's proxy headers yet?

File: libc/test/integration/src/pthread/pthread_spinlock_test.cpp [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+#include "src/pthread/pthread_create.h"
+#include "src/pthread/pthread_join.h"
+#include "src/pthread/pthread_spin_destroy.h"
+#include "src/pthread/pthread_spin_init.h"
+#include "src/pthread/pthread_spin_lock.h"
+#include "src/pthread/pthread_spin_trylock.h"
+#include "src/pthread/pthread_spin_unlock.h"
+#include "test/IntegrationTest/test.h"
+#include <errno.h>
```
Comment (lntue) [inline libc/test/integration/src/pthread/pthread_spinlock_test.cpp:17]: Use proxy header for errno macros.

File: libc/src/pthread/pthread_spin_destroy.h (hunk 3 above)
Comment (SchrodingerZhu) [reply libc/src/pthread/pthread_spin_destroy.h:13]: I guess no. Such routines are hardly used in overlay mode. That is being said, I think we should open one as many implementations do support overlay mode.

File: libc/test/integration/src/pthread/pthread_spinlock_test.cpp (hunk 4 above)
Comment (SchrodingerZhu) [reply libc/test/integration/src/pthread/pthread_spinlock_test.cpp:17]: Nice catch

**Review by lntue — APPROVED** [review]
Thanks!  This is much clearer than the previous version!

**Pull request merged**